src/main/java/io/pravega/connectors/flink/source/enumerator/PravegaSplitEnumerator.java

```diff
@@ -0,0 +1,200 @@
+                    .groupRefreshTimeMillis(this.readerGroupConfig.getGroupRefreshTimeMillis())
+                    .disableAutomaticCheckpoints()
+                    .startFromCheckpoint(this.checkpoint)
+                    .build());
+        }
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+        // the Pravega source pushes splits eagerly, rather than act upon split requests
+    }
+
+    // check with the current parallelism and update the assignment (always one-to-one mapping)
+    @Override
+    public void addReader(int subtaskId) {
+        if (enumContext.registeredReaders().size() == enumContext.currentParallelism() && splits.size() > 0) {
+            LOG.debug("Adding reader {} to PravegaSplitEnumerator for reader group: {}/{}.",
+                    subtaskId,
+                    scope,
+                    readerGroupName);
+            int numReaders = enumContext.registeredReaders().size();
+            Map<Integer, List<PravegaSplit>> assignment = new HashMap<>();
+            for (int i = 0; i < numReaders; i++) {
+                assignment.put(i, Collections.singletonList(splits.get(i)));
+            }
+            enumContext.assignSplits(new SplitsAssignment<>(assignment));
+            splits.clear();
+        }
+    }
+
+    // deal with the data recovery, it will call readerGroup::initiateCheckpoint to get a checkpoint for recovery.
+    // this call will be handled in another thread pool instead of the split enumerator thread.
+    @Override
+    public Checkpoint snapshotState(long chkPtID) throws Exception {
+        final String checkpointName = createCheckpointName(chkPtID);
+
+        LOG.info("Initiate checkpoint {}", checkpointName);
+        final CompletableFuture<Checkpoint> checkpointResult =
+                this.readerGroup.initiateCheckpoint(checkpointName, scheduledExecutorService);
+        try {
+            this.checkpoint = checkpointResult.get(5, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            LOG.error("Pravega checkpoint met error.", e);
+            this.checkpoint = null;
+        }
+
+        return checkpoint;
+    }
+
+    @Override
+    public void close() throws IOException {
+        readerGroup.close();
+        readerGroupManager.close();
+    }
+
+    // This method is called when it tries to "regionally" recover from a failure to assign unassigned splits. We will
+    // throw an intentional exception here to trigger a global recovery since Pravega since Pravega
+    // doesn't have a mechanism to recover from partial failure. This will shutdown the reader group,
+    // recreate it and recover it from the latest checkpoint.
+    @Override
+    public void addSplitsBack(List<PravegaSplit> splits, int subtaskId) {
+        LOG.info("Call addSplitsBack {} : {}", splits.size(), subtaskId);
+
+        if (!isRecovered) {
```

Comment:
yes and no. `addSplitsBack` is called as the local recover logic and it won't deal with checkpoint. Currently one source reader is mapped to one split reader(Event Stream Reader). If it fails we have to reset the reader group from the checkpoint as Pravega's recovery mechanism. Hence we throw an exception here to trigger a global failure to reset the split enumerator since reader group is inside enumerator's scope.

Reply:
I'm not sure I understand why it needs to reset to a checkpoint. Could we make the event reader offline instead? Perhaps I'm not understanding what is expected to fail here.